File: app/src/main/java/life/plank/juna/zone/view/adapter/EmojiAdapter.java
package life.plank.juna.zone.view.adapter;

import android.content.Context;
import android.support.design.widget.BottomSheetBehavior;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.google.gson.JsonObject;

import java.net.HttpURLConnection;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.network.interfaces.RestApi;
import retrofit2.Response;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static life.plank.juna.zone.ZoneApplication.getApplication;
import static life.plank.juna.zone.util.DateUtil.getRequestDateStringOfNow;
import static life.plank.juna.zone.util.PreferenceManager.getToken;
import static life.plank.juna.zone.util.RestUtilKt.errorToast;
import static life.plank.juna.zone.util.UIDisplayUtil.emoji;

public class EmojiAdapter extends RecyclerView.Adapter<EmojiAdapter.EmojiViewHolder> {

    private static final String TAG = EmojiAdapter.class.getSimpleName();
    public static String feedId;
    public BottomSheetBehavior emojiBottomSheetBehavior;
    @Inject
    @Named("default")
    RestApi restApi;
    private Context context;
    private String boardId;

    public EmojiAdapter(Context activity, String boardId, BottomSheetBehavior emojiBottomSheetBehavior) {
        this.context = activity;
        this.boardId = boardId;
        this.emojiBottomSheetBehavior = emojiBottomSheetBehavior;
    }

    @Override
    public EmojiViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        getApplication().getUiComponent().inject(this);
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_emoji, parent, false);
        return new EmojiViewHolder(view);
    }

    @Override
    public void onBindViewHolder(EmojiViewHolder holder, int position) {
        holder.emojiView.setImageDrawable(context.getDrawable(emoji[position].getEmojiUrl()));
        holder.emojiView.setOnClickListener(view -> {
            postEmoji(feedId, boardId, emoji[position].getReaction());
        });
    }

    private void postEmoji(String feedItemId, String boardId, Integer reaction) {
        restApi.postReaction(feedItemId, boardId,  reaction, getRequestDateStringOfNow(), getToken())

                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<Response<JsonObject>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted: ");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, "onError: " + e);
                        errorToast(R.string.something_went_wrong, e);
                    }

                    @Override
                    public void onNext(Response<JsonObject> response) {
                        switch (response.code()) {
                            case HttpURLConnection.HTTP_CREATED:
                                emojiBottomSheetBehavior.setPeekHeight(0);
                                emojiBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
                                break;
                            default:
                                errorToast(R.string.something_went_wrong, response);
                                break;
                        }
                    }
                });
    }

    @Override
    public int getItemCount() {
        return emoji.length;
    }

    static class EmojiViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.emoji_view)
        ImageView emojiView;

        EmojiViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/view/adapter/EmojiAdapter.kt
package life.plank.juna.zone.view.adapter

import android.support.design.widget.BottomSheetBehavior
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.item_emoji.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.ZoneApplication
import life.plank.juna.zone.data.network.interfaces.RestApi
import life.plank.juna.zone.util.DateUtil.getRequestDateStringOfNow
import life.plank.juna.zone.util.PreferenceManager.Auth.getToken
import life.plank.juna.zone.util.UIDisplayUtil.emoji
import life.plank.juna.zone.util.errorToast
import life.plank.juna.zone.util.facilis.hide
import life.plank.juna.zone.util.facilis.onDebouncingClick
import life.plank.juna.zone.util.setObserverThreadsAndSmartSubscribe
import java.net.HttpURLConnection.HTTP_OK

class EmojiAdapter(
        private val restApi: RestApi,
        private val boardId: String,
        private val emojiBottomSheetBehavior: BottomSheetBehavior<*>?
) : RecyclerView.Adapter<EmojiAdapter.EmojiViewHolder>() {

    companion object {
        var feedId: String? = null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EmojiViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_emoji, parent, false)
        return EmojiViewHolder(view)
    }

    override fun onBindViewHolder(holder: EmojiViewHolder, position: Int) {
        holder.itemView.emoji_view.setImageDrawable(ZoneApplication.getContext().getDrawable(emoji[position].emojiDrawable!!))
        holder.itemView.emoji_view.onDebouncingClick {
            emojiBottomSheetBehavior?.hide()
            postEmoji(emoji[position].reaction)
        }
    }

    private fun postEmoji(reaction: Int?) {
        restApi.postReaction(feedId, boardId, reaction, getRequestDateStringOfNow(), getToken()).setObserverThreadsAndSmartSubscribe({
            Log.e("postEmoji()", "onError: ", it)
        }, {
            if (it.code() != HTTP_OK) errorToast(R.string.something_went_wrong, it)
        })
    }

    override fun getItemCount(): Int = emoji.size

    class EmojiViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}
